### PR TITLE
[Enhance] Speed up the inference by reducing numpy array copy and doing Normalization in GPU

### LINKED ISSUE #1316: Speedup the 'TopDownAffine' by give short cut to rotation=0
The `TopDownAffine` is a **disaster** for cpu load. Because in most case of inference, there is no ROTATION at all. Please give a shortcut when rotation==0. The cv2.warpAffine is so laborious.
```
# mmpose/datasets/pipelines/top_down_transform.py
@PIPELINES.register_module()
class TopDownAffine:
      ...
        c = results['center']
        s = results['scale']
        r = results['rotation']
         ....
        img = [
                    cv2.warpAffine(
                        i,
                        trans, (int(image_size[0]), int(image_size[1])),
                        flags=cv2.INTER_LINEAR) for i in img
                ]
```

the configure file 
```
#configs/wholebody/2d_kpt_sview_rgb_img/topdown_heatmap/coco-wholebody/res152_coco_wholebody_384x288.py
val_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(type='TopDownAffine'),
    dict(type='ToTensor'),
    dict(
        type='NormalizeTensor',
        mean=[0.485, 0.456, 0.406],
        std=[0.229, 0.224, 0.225]),
    dict(
        type='Collect',
        keys=['img'],
        meta_keys=[
            'image_file', 'center', 'scale', 'rotation', 'bbox_score',
            'flip_pairs'
        ]),
]
```
I used GPU to do inference of a video by `demo/top_down_video_demo_full_frame_without_det.py`. However, the cpu workload is surprising 2000%. Part of the result own to that I used 1280x800 resolution video and 512x512 feature size, but the most because of this laborious `TopDownAffine`.

### PR DESCRIPTION
**Pull request by chenxinfeng4**
Allow ToTensor to automaticly upload data to gpu, so that Normalize will be done in GPU.
Fix #1316

## Motivation

The inference for big image size is so slow. Because there are so many numpy array coping. And the data normalize should also be done in GPU if possible. After my revision, the cpu usage is significantly reduced, and the speed is increased.

## Checklist

**Before PR**:

- [x] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) are used to fix the potential lint issues.
- [x] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] New functionalities are covered by complete unit tests. If not, please add more unit tests to ensure correctness.
- [x] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmpose/datasets/pipelines/loading.py [hunk 1]
```diff
@@ -71,10 +71,10 @@ def __call__(self, results):
                 # The original results['img'] is assumed to be image(s) in BGR
                 # order, so we convert the color according to the arguments.
                 if results['img'].ndim == 3:
-                    results['img'] = mmcv.bgr2rgb(results['img'])
+                    results['img'] = results['img'][:, :, ::-1]
                 elif results['img'].ndim == 4:
                     results['img'] = np.concatenate(
-                        [mmcv.bgr2rgb(img) for img in results['img']], axis=0)
+                        [results['img'][:, :, ::-1] for img in results['img']], axis=0)
```
Comment (jin-s13) [inline mmpose/datasets/pipelines/loading.py:77]: should be `[img[:, :, ::-1] for img in results['img']]` ?

Comment (chenxinfeng4) [reply mmpose/datasets/pipelines/loading.py:77]: I'll fix it

File: mmpose/datasets/pipelines/shared_transform.py [hunk 2]
```diff
@@ -27,11 +28,18 @@ class ToTensor:
...
+        to_tensor = (lambda x: torch.from_numpy(x.astype('float32')).permute(
+            2, 0, 1).to(self.device))
         if isinstance(results['img'], (list, tuple)):
-            results['img'] = [F.to_tensor(img) for img in results['img']]
+            results['img'] = [
+                to_tensor(img.astype('float32')) for img in results['img']
+            ]
         else:
-            results['img'] = F.to_tensor(results['img'])
+            results['img'] = to_tensor(results['img'].astype('float32'))
```
Comment (jin-s13) [inline mmpose/datasets/pipelines/shared_transform.py:42]: It seems that `.astype('float32')` is not necessary, because in `to_tensor`, x will first transformed to 'float32'.

File: mmpose/datasets/pipelines/loading.py (hunk 1 above)
Comment (chenxinfeng4) [reply mmpose/datasets/pipelines/loading.py]: I use `img[:, :, ::-1]` instead of `bgr2rgb(img)`, because the former is ligthweight for cpu.

File: mmpose/datasets/pipelines/shared_transform.py (hunk 2 above)
Comment (chenxinfeng4) [reply mmpose/datasets/pipelines/shared_transform.py:42]: I'll fix it

**Comment by jin-s13** [comment]
Thanks for contributing! @chenxinfeng4 
This PR has some lint issues. Please follow the instructions at https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md to setup the pre-commit.

**Comment by chenxinfeng4** [comment]
I've tried the lint formating. Can you check it?

**Review by liqikai9 — APPROVED** [review]

**Comment by chenxinfeng4** [comment]
image
This is test on `demo/top_down_video_demo_full_frame_without_det.py` with image_size=512x512. The FPS is a little bit slower (because the `img[:, :, ::-1]` for bgr2rgb is slower than `cv2.cvtColor`, but the former is cpu friendly.

The CPU workload was once the bottleneck in multiple GPUs inference in the origin case.

**Comment by chenxinfeng4** [comment]
I'll improve the `demo/top_down_video_demo_full_frame_without_det.py` to 130fps with even less cpu workload （400%）.  CPU。这将是我下一个PR。
image

**Comment by liqikai9** [comment]
Your improvement looks great! Do you have corresponding speed test results running on the GPU as well?

**Comment by chenxinfeng4** [comment]
> 

Yeah. They were the tested in GPU NVIDIA3090. Beacuse both cases didn't show difference in GPU work load (arround 40-60% flucturating), so I didn't pose the GPU usage.

File: mmpose/datasets/pipelines/shared_transform.py [hunk 3]
```diff
@@ -27,11 +28,16 @@ class ToTensor:
         results (dict): contain all information about training.
     """
 
+    def __init__(self, device='cpu'):
+        self.device = device
+
     def __call__(self, results):
+        to_tensor = (lambda x: torch.from_numpy(x.astype('float32')).permute(
```
Comment (liqikai9) [inline mmpose/datasets/pipelines/shared_transform.py:35]: It seems that when I used **multiple GPUs** and this [config](https://github.com/open-mmlab/mmpose/blob/master/configs/body/2d_kpt_sview_rgb_img/topdown_heatmap/coco/hrnet_w32_coco_256x192.py) for **testing**, the AP is near **zero**, which is not reasonable.

But when I used **single GPU**, the result is consistent . 

I also tried theses two cases:
1. Only keep your modification about in the `ToTensor`, the result with single GPU is correct while result in multiple GPUs is still near **zero**. 
2. Only keep your modification about in the `NormalizeTensor`, the results on single and multiple GPUa are both correct. 

I was wondering if it have something to do with the modifications you add inside `ToTensor`, which may cause some in place modification to the data implicitly.

Comment (chenxinfeng4) [reply mmpose/datasets/pipelines/shared_transform.py:35]: I didn't test in multiple GPU before. It seems unreasonable that the modification won't work in multiple GPU. What's the test file and command did you use? I will then check it and find the way out.

Comment (liqikai9) [reply mmpose/datasets/pipelines/shared_transform.py:35]: **UPDATE**: Previously I mistook the running path of MMPose. Now the results of single GPU and multiple GPUs are the same, but both got mAP near **zero**.

I use this config https://github.com/open-mmlab/mmpose/blob/master/configs/body/2d_kpt_sview_rgb_img/topdown_heatmap/coco/hrnet_w32_coco_256x192.py.  and download the corresponding checkpoint from this link https://download.openmmlab.com/mmpose/top_down/hrnet/hrnet_w32_coco_256x192-c78dce93_20200708.pth

1. For single GPU testing, I run:
```
PYTHONPATH="$(dirname $0)/..":$PYTHONPATH python tools/test.py configs/body/2d_kpt_sview_rgb_img/topdown_heatmap/coco/hrnet_w32_coco_256x192.py checkpoints/hrnet_w32_coco_256x192-c78dce93_20200708.pth
```

2. For multiple GPUs testing, I run:
```
GPUS=8 GPUS_PER_NODE=8 CPUS_PER_TASK=12 SRUN_ARGS='--quotatype=auto' ./tools/slurm_test.sh mm_human test_pr1320 configs/body/2d_kpt_sview_rgb_img/topdown_heatmap/coco/hrnet_w32_coco_256x192.py checkpoints/hrnet_w32_coco_256x192-c78dce93_20200708.pth
```

Another interesting phenomena is that when I run multi-GPU training using the same config and this command:
```
GPUS=8 GPUS_PER_NODE=8 CPUS_PER_TASK=12 SRUN_ARGS='--quotatype=auto' ./tools/slurm_train.sh mm_human train_pr1320 configs/body/2d_kpt_sview_rgb_img/topdown_heatmap/coco/hrnet_w32_coco_256x192.py work_dirs/
```
The evaluation results during training **looked just fine**., with mAP near 0.65 after 10 epochs. 

Considering the above testing results, this is a little bit strange.

Comment (liqikai9) [reply mmpose/datasets/pipelines/shared_transform.py:35]: When I change your modification in `ToTensor` back to `F.to_tensor`, the mAP is correct. 

So maybe it was the `to_tensor` function you add that somehow changes the data inplace.

**Comment by chenxinfeng4** [comment]
Could that be the `inplace` in normalize?
```
results['img'] = F.normalize(
                results['img'], mean=self.mean, std=self.std, inplace=True)
```

I don't think the `to_tensor` has the issue. The `to_tensor` should copy the numpy to GPU without modification of the original numpy.
```
to_tensor = (lambda x: torch.from_numpy(x.astype('float32')).permute(
            2, 0, 1).to(self.device))
```

**Comment by liqikai9** [comment]
I have tried these experiments.

1. I change 
```
results['img'] = F.normalize(
                results['img'], mean=self.mean, std=self.std, inplace=True)
```
to 

```
results['img'] = F.normalize(
                results['img'], mean=self.mean, std=self.std)
```
and keep your `to_tensor` function, the mAP is near zero.

2. I change your to_tensor function back to the previous way:
`results['img'] = F.to_tensor(results['img'])`
and keep your modification in normalize:
```
results['img'] = F.normalize(
                results['img'], mean=self.mean, std=self.std, inplace=True)
```
then the mAP is correct.

**Comment by liqikai9** [comment]
Update: **I think I have found the problem.**

At first, the img array read after the pipeline `LoadImageFromFile` is in the range of **[0, 255]**.

It seems that in `F.to_tensor ` (check the doc [here](https://pytorch.org/vision/stable/generated/torchvision.transforms.ToTensor.html#torchvision.transforms.ToTensor))  the img array will be converted to the tensor in the range **[0.0, 1.0]** .
However, your `to_tensor` function just change the img array to tensor but didn't converted them to the range **[0.0, 1.0]**. 

I tried to change your to_tensor to:
```
to_tensor = (lambda x: torch.from_numpy(x.astype('float32')).permute(
            2, 0, 1).to(self.device)) / 255.0
```
without changing other places. The mAP is correct!!

Also, this can explain why during training, the mAP is near 0.7. I think during the training, the model can learn how to scale the img so the result is not zero.

**Comment by liqikai9** [comment]
You may change your `to_tensor` function, and see check the speed test result on CPU.

**Comment by chenxinfeng4** [comment]
Ohhh, it surprised me for the difference of `F.to_tensor` and `to_tensor`. I'll update that and check the speed again.

**Comment by chenxinfeng4** [comment]
my to_tensor(),  65.35it/s
image

F.to_tensor(), 61.98it/s
image

File: mmpose/datasets/pipelines/shared_transform.py [hunk 4]
```diff
@@ -27,11 +28,16 @@ class ToTensor:
...
+            2, 0, 1).to(self.device) / 255.0)
```
Comment (liqikai9) [inline mmpose/datasets/pipelines/shared_transform.py:36]: The `255.0` is a little bit hard-coded here. Is this always necessary? Because in the source code of `F.to_tensor`, 

https://github.com/pytorch/vision/blob/e9fb2a356384180d92dc60c49b4dd4b5b3c6ee8f/torchvision/transforms/functional.py#L152-L155

it seems that if the img is already located in [0.0, 1.0], it just returns the img tensor itself. 

BTW, `F.to_tensor` uses `div(255)`, I am not sure if this is faster. You may have a try!

Comment (chenxinfeng4) [reply mmpose/datasets/pipelines/shared_transform.py:36]: I've tested that in GPU and found the `division of 255.0` makes no difference in speed. The `division` is just similar to Normalization, which can also be speed up in GPU.

**Comment by chenxinfeng4** [comment]
Why the merging is blocked? How can I fix it?

**Comment by ly015** [comment]
> Why the merging is blocked? How can I fix it?

We are double-checking the optimized performance on our local devices. This feature will be merged soon and included in the next mmpose release.

File: mmpose/datasets/pipelines/loading.py [hunk 5]
```diff
@@ -71,10 +71,10 @@ def __call__(self, results):
                 # The original results['img'] is assumed to be image(s) in BGR
                 # order, so we convert the color according to the arguments.
                 if results['img'].ndim == 3:
-                    results['img'] = mmcv.bgr2rgb(results['img'])
```
Comment (ly015) [inline mmpose/datasets/pipelines/loading.py:74]: Numpy tensor indexing is faster than `mmcv.bgr2rgb` but the cost is 1e-6 sec/image level, which should be negligible. So I think we don't need to change the original code since it's more readable.

File: mmpose/datasets/pipelines/shared_transform.py (hunk 4 above)
Comment (ly015) [reply mmpose/datasets/pipelines/shared_transform.py:36]: I suggest using the in-place operation `dev_(255)` to save memory.

**Review by liqikai9 — APPROVED** [review]
